### PR TITLE
refactor: stop experiment modal [DET-6325]

### PR DESCRIPTION
**Pull request by hkang1**
## Description

With more modal use, refactor a few experiment related modals to:
1. use a hook pattern instead of defining a React component
2. build and use a shared modal hook for basic modal functionality
3. fix the UI blocking behavior for modals that have a Promise based operation upon modal submission

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/pages/ExperimentDetails/ExperimentDetailsHeader.tsx [hunk 1]
```diff
@@ -255,7 +307,38 @@ const ExperimentDetailsHeader: React.FC<Props> = ({
         leftContent={(
           <div className={css.base}>
             <div className={css.experimentInfo}>
-              <ExperimentState experiment={experiment} />
+              <Spinner spinning={isChangingState}>
+                <div className={css.experimentState} style={{ backgroundColor }}>
+                  {experiment.state === RunState.Active && (
```
Comment (hamidzr) [inline webui/react/src/pages/ExperimentDetails/ExperimentDetailsHeader.tsx]: question: do we need to consider job states? maybe using `pausableStates`

File: webui/react/src/pages/ExperimentDetails/ExperimentDetailsHeader.tsx [hunk 2]
```diff
@@ -255,7 +307,38 @@ const ExperimentDetailsHeader: React.FC<Props> = ({
...
+                  {experiment.state === RunState.Paused && (
+                    <Button
+                      className={css.buttonPlay}
+                      ghost={true}
+                      icon={<Icon name="play" size="large" />}
+                      shape="circle"
+                      onClick={handlePlayClick}
+                    />
+                  )}
+                  {[ RunState.Active, RunState.Paused ].includes(experiment.state) && (
```
Comment (hamidzr) [inline webui/react/src/pages/ExperimentDetails/ExperimentDetailsHeader.tsx]: ditto

File: webui/react/src/hooks/useModal/useModal.tsx [hunk 3]
```diff
@@ -9,11 +9,16 @@ import { isAsyncFunction } from 'utils/data';
...
 
 export interface ModalHooks {
-  modalClose: () => void;
+  modalClose: (reason?: ModalCloseReason) => void;
   modalOpen: (modalProps?: ModalFuncProps) => void;
   modalRef: React.MutableRefObject<ReturnType<ModalFunc> | undefined>;
 }
 
+interface ModalOptions {
+  rawCancel?: boolean;
```
Comment (hamidzr) [inline webui/react/src/hooks/useModal/useModal.tsx:18]: suggestion: add comments for the options

**Review by hamidzr — APPROVED** [review]
LGTM.

nit suggestion: this probably would have worked well as a smaller initial pr to introduce the useModal hook with just addressing say stop experiment modal. Theoretically, it could have saved work if we were to update the design.

File: webui/react/src/pages/ExperimentDetails/ExperimentDetailsHeader.tsx (hunk 1 above)
Comment (hkang1) [reply webui/react/src/pages/ExperimentDetails/ExperimentDetailsHeader.tsx]: ah yeah I added them when I rebased with your job states work.

**Pull request merged**